packages/amesos2/src/Amesos2_Util.hpp

```diff
@@ -1023,6 +1027,166 @@ namespace Amesos2 {
+            if (i != host_cols(j)) {
+              host_strip_diag_cols(new_nnz++) = host_cols(j);
+            }
+          }
+        }
+        host_strip_diag_row_ptr(size) = new_nnz;
+
+        // we'll get original permutations on host
+        host_metis_array host_perm(
+          Kokkos::ViewAllocateWithoutInitializing("host_perm"), size);
+        host_metis_array host_peri(
+          Kokkos::ViewAllocateWithoutInitializing("host_peri"), size);
+
+        // If we want to remove metis.h included in this header we can move this
+        // to the cpp, but we need to decide how to handle the idx_t declaration.
+        idx_t metis_size = size;
+        int err = METIS_NodeND(&metis_size, host_strip_diag_row_ptr.data(), host_strip_diag_cols.data(),
+          NULL, NULL, host_perm.data(), host_peri.data());
+
+        TEUCHOS_TEST_FOR_EXCEPTION(err != METIS_OK, std::runtime_error,
+          "METIS_NodeND failed to sort matrix.");
+
+        // exec space to resort the matrix
+        // right now we're assuming all three vectors are in fact on the same
+        // memory space but this is not necessarily true for future use. Then
+        // we'll need to make the exec space an option to set and they'll all
+        // need to be mirrored into that space. Add a compilation failure here
+        // to make that clear to a future user.
+        typedef typename values_view_t::execution_space exec_space_t;
+        typedef typename cols_view_t::execution_space cols_exec_space_t;
+        typedef typename row_ptr_view_t::execution_space row_ptr_exec_space_t;
+        static_assert(std::is_same<exec_space_t, cols_exec_space_t>::value &&
+                      std::is_same<cols_exec_space_t, row_ptr_exec_space_t>::value,
+                      "Amesos2 reorder method is currently assuming all three "
+                      "matrix vectors are on the same memory space. If this "
+                      "requirement is changing it will need some minor updates "
+                      "to make it work with differing memory spaces.");
+
+        // put the permutations on our saved device ptrs
+        // these will be used to permute x and b when we solve
+        auto device_perm = Kokkos::create_mirror_view(exec_space_t(), host_perm);
+        auto device_peri = Kokkos::create_mirror_view(exec_space_t(), host_peri);
+        deep_copy(device_perm, host_perm);
+        deep_copy(device_peri, host_peri);
+
+        // we'll permute matrix on device to a new set of arrays
+        row_ptr_view_t new_row_ptr(
+          Kokkos::ViewAllocateWithoutInitializing("new_row_ptr"), row_ptr.size());
+        cols_view_t new_cols(
+          Kokkos::ViewAllocateWithoutInitializing("new_cols"), cols.size() - new_nnz/2);
+        values_view_t new_values(
+          Kokkos::ViewAllocateWithoutInitializing("new_values"), values.size() - new_nnz/2);
+
+        // permute row indices
+        Kokkos::RangePolicy<exec_space_t> policy_row(0, row_ptr.size());
```

Comment:
@MicheldeMessieres I think a fence may need to be added before the `parallel_scan` as a precaution to prevent issues when UVM is enabled to ensure data is properly synced (unless all device Views are explicitly templated on `Kokkos::CudaSpace` when Cuda is enabled).

Reply:
Currently cuSolver will work explicitly on CudaSpace or Host but not UVM. The Kokkos adapter tests will send all 3 types (Host, CudaUVMSpace, CudaSpace) as inputs but the solver itself is never using UVM views. For Cholmod the views are always Host. For Tacho, the Views are still UVMSpace due to the TaskScheduler issues we had. Tacho doesn't use this ordering path and we plan to make it also just CudaSpace once the other issues are fixed. So perhaps the fence is not needed presently. Not sure if we want to start anticipating CUDA_LAUNCH_BLOCKING off formatting yet.